File: 0396-rotate-function/0396-rotate-function.cpp
class Solution {
public:
    int maxRotateFunction(vector<int>& nums) {
        int n = nums.size();
        long long total=0, f=0;

        for(int i=0; i<n; i++){
            total+=nums[i];
            f+=(long long)i*nums[i];
        }

        long long maxf=f;

        for(int k=1; k<n; k++){
            f=f+total - (long long)n * nums[n-k];
            maxf=max(maxf,f);
        }
        return (int)maxf;
    }
};

File: 0396-rotate-function/0396-rotate-function.py
class Solution(object):
    def maxRotateFunction(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        n=len(nums)
        total=sum(nums)

        f=sum(i*nums[i] for i in range(n))
        maxf=f

        for k in range(1,n):
            f=f+total-n*nums[n-k]
            maxf=max(maxf,f)

        return maxf